File: Sources/CodyUsageCore/UsageCoordinator.swift
import Foundation

public final class UsageCoordinator: @unchecked Sendable {
    public typealias UpdateHandler = @Sendable (UsageSnapshot) -> Void

    private let client: AppServerClient
    private let rolloutReader: RolloutReader
    private let eventMonitor: SessionEventMonitor
    private let queue = DispatchQueue(label: "CodyUsageOverlay.Coordinator")
    private var timer: DispatchSourceTimer?
    private var snapshot = UsageSnapshot()
    private var updateHandler: UpdateHandler?
    private var retrySeconds: TimeInterval = 2
    private var retryWorkItem: DispatchWorkItem?
    private var rateLimitRefreshInFlight = false
    private var rateLimitRefreshPending = false
    private var lastRateLimitUpdate: Date?
    private var lastContextUpdate: Date?

    public init(client: AppServerClient = AppServerClient(), rolloutReader: RolloutReader = RolloutReader(), eventMonitor: SessionEventMonitor = SessionEventMonitor()) {
        self.client = client
        self.rolloutReader = rolloutReader
        self.eventMonitor = eventMonitor
        client.setNotificationHandler { [weak self] method, params in
            guard method == "account/rateLimits/updated" else { return }
            self?.applyRateLimits(params)
        }
    }

    public func onUpdate(_ handler: @escaping UpdateHandler) { queue.async { self.updateHandler = handler } }

    public func start() {
        queue.async { self.refreshContext() }
        eventMonitor.start { [weak self] paths in
            guard let self else { return }
            self.queue.async { self.refreshContext(changedPaths: paths) }
        }
        queue.async {
            guard self.timer == nil else { return }
            let timer = DispatchSource.makeTimerSource(queue: self.queue)
            timer.schedule(deadline: .now(), repeating: 1)
            timer.setEventHandler { [weak self] in self?.tick() }
            self.timer = timer
            timer.resume()
        }
        refreshRateLimits()
    }

    public func stop() {
        eventMonitor.stop()
        queue.async {
            self.timer?.cancel()
            self.timer = nil
            self.retryWorkItem?.cancel()
            self.retryWorkItem = nil
            self.rateLimitRefreshPending = false
        }
        Task { try? await client.stop() }
    }

    public func forceRefresh() { refreshRateLimits(force: true); queue.async { self.refreshContext() } }

    public func diagnostics() -> String {
        queue.sync {
            let executable = AppServerClient.locateExecutable()?.path ?? "missing"
            return "Codex: \(executable)\nThread: \(snapshot.activeThreadId ?? "unknown")\nFreshness: \(snapshot.freshness.rawValue)\nUpdated: \(snapshot.lastUpdatedAt)"
        }
    }

    private func tick() {
        if lastRateLimitUpdate == nil || Date().timeIntervalSince(lastRateLimitUpdate!) >= 60 { refreshRateLimits() }
        let newest = [lastRateLimitUpdate, lastContextUpdate].compactMap { $0 }.max() ?? .distantPast
        if Date().timeIntervalSince(newest) > 90 && snapshot.freshness == .fresh {
            snapshot.freshness = .delayed; publish()
        }
    }

    private func refreshContext() {
        if let usage = rolloutReader.refresh() {
            applyContext(usage)
        }
    }

    private func refreshContext(changedPaths: [String]) {
        if let usage = rolloutReader.refresh(changedPaths: changedPaths) {
            applyContext(usage)
        }
    }

    private func applyContext(_ usage: ContextUsage) {
            snapshot.contextRemainingPercent = usage.remainingPercent
            snapshot.activeThreadId = usage.threadId
            if let rateLimits = usage.rateLimits { mergeRateLimits(rateLimits) }
            lastContextUpdate = usage.updatedAt
            snapshot.lastUpdatedAt = Date()
            snapshot.freshness = .fresh
            publish()
    }

    private func refreshRateLimits(force: Bool = false) {
        queue.async { self.beginRateLimitRefresh(force: force) }
    }

    private func beginRateLimitRefresh(force: Bool) {
        if force {
            retryWorkItem?.cancel()
            retryWorkItem = nil
        } else if retryWorkItem != nil {
            return
        }
        guard !rateLimitRefreshInFlight else {
            if force { rateLimitRefreshPending = true }
            return
        }
        rateLimitRefreshInFlight = true
        Task {
            do {
                let raw = try await client.readRateLimits()
                queue.async {
                    self.rateLimitRefreshInFlight = false
                    self.retrySeconds = 2
                    self.applyRateLimitsOnQueue(raw)
                    self.runPendingRateLimitRefreshIfNeeded()
                }
            } catch {
                queue.async {
                    self.rateLimitRefreshInFlight = false
                    let hasLimits = self.snapshot.fiveHourRemainingPercent != nil || self.snapshot.weeklyRemainingPercent != nil
                    self.snapshot.freshness = hasLimits ? .delayed : .unavailable
                    self.publish()
                    if self.runPendingRateLimitRefreshIfNeeded() { return }
                    let delay = self.retrySeconds
                    self.retrySeconds = min(60, self.retrySeconds * 2)
                    let retry = DispatchWorkItem { [weak self] in
                        guard let self else { return }
                        self.retryWorkItem = nil
                        self.beginRateLimitRefresh(force: false)
                    }
                    self.retryWorkItem = retry
                    self.queue.asyncAfter(deadline: .now() + delay, execute: retry)
                }
            }
        }
    }

    @discardableResult
    private func runPendingRateLimitRefreshIfNeeded() -> Bool {
        guard rateLimitRefreshPending else { return false }
        rateLimitRefreshPending = false
        beginRateLimitRefresh(force: true)
        return true
    }

    private func applyRateLimits(_ raw: Data) {
        queue.async { self.applyRateLimitsOnQueue(raw) }
    }

    private func applyRateLimitsOnQueue(_ raw: Data) {
        do {
            let result = try RateLimitParser.parse(data: raw)
            mergeRateLimits(result)
            lastRateLimitUpdate = Date()
            snapshot.lastUpdatedAt = Date()
            snapshot.freshness = .fresh
        } catch {
            snapshot.freshness = .incompatible
        }
        publish()
    }

    private func mergeRateLimits(_ result: RateLimitResult) {
        if let five = result.fiveHour {
            snapshot.fiveHourRemainingPercent = mergedRemaining(
                current: snapshot.fiveHourRemainingPercent,
                currentReset: snapshot.fiveHourResetsAt,
                incoming: five
            )
            snapshot.fiveHourResetsAt = five.resetsAt
        }
        if let weekly = result.weekly {
            snapshot.weeklyRemainingPercent = mergedRemaining(
                current: snapshot.weeklyRemainingPercent,
                currentReset: snapshot.weeklyResetsAt,
                incoming: weekly
            )
            snapshot.weeklyResetsAt = weekly.resetsAt
        }
    }

    private func mergedRemaining(current: Int?, currentReset: Date?, incoming: RateLimitWindow) -> Int {
        guard let current,
              let currentReset,
              let incomingReset = incoming.resetsAt,
              abs(currentReset.timeIntervalSince(incomingReset)) < 1 else {
            return incoming.remainingPercent
        }
        return min(current, incoming.remainingPercent)
    }

    private func publish() { updateHandler?(snapshot) }
}

File: windows/src/CodyUsageOverlay.Core/UsageCoordinator.cs
namespace CodyUsageOverlay.Core;

public sealed class UsageCoordinator(IAppServerClient client, ISessionMonitor sessionMonitor) : IAsyncDisposable
{
    private readonly object gate = new();
    private CancellationTokenSource? lifetime;
    private UsageSnapshot snapshot = new(LastUpdatedAt: DateTimeOffset.Now);
    private DateTimeOffset? lastRateLimitUpdate;
    private DateTimeOffset? lastContextUpdate;
    public event Action<UsageSnapshot>? Updated;
    public UsageSnapshot Snapshot { get { lock (gate) return snapshot; } }

    public void Start()
    {
        if (lifetime is not null) return;
        lifetime = new CancellationTokenSource();
        client.Notification += OnNotification;
        sessionMonitor.Updated += OnContext;
        sessionMonitor.Start();
        _ = RefreshLoopAsync(lifetime.Token);
        _ = FreshnessLoopAsync(lifetime.Token);
    }

    public Task ForceRefreshAsync(CancellationToken cancellationToken = default) => RefreshRateLimitsAsync(cancellationToken);

    private async Task RefreshLoopAsync(CancellationToken cancellationToken)
    {
        var retry = TimeSpan.FromSeconds(2);
        while (!cancellationToken.IsCancellationRequested)
        {
            try
            {
                await RefreshRateLimitsAsync(cancellationToken);
                retry = TimeSpan.FromSeconds(60);
            }
            catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested) { break; }
            catch
            {
                lock (gate) snapshot = snapshot with { Freshness = HasAnyValue(snapshot) ? Freshness.Delayed : Freshness.Unavailable };
                Publish();
                retry = TimeSpan.FromSeconds(Math.Min(60, Math.Max(2, retry.TotalSeconds * 2)));
            }
            try { await Task.Delay(retry, cancellationToken); } catch (OperationCanceledException) { break; }
        }
    }

    private async Task RefreshRateLimitsAsync(CancellationToken cancellationToken)
    {
        var raw = await client.ReadRateLimitsAsync(cancellationToken);
        Merge(RateLimitParser.Parse(raw));
    }

    private void OnNotification(string method, byte[] payload)
    {
        if (!method.Equals("account/rateLimits/updated", StringComparison.OrdinalIgnoreCase)) return;
        try { Merge(RateLimitParser.Parse(payload)); }
        catch { lock (gate) snapshot = snapshot with { Freshness = Freshness.Incompatible }; Publish(); }
    }

    private void Merge(RateLimitResult result)
    {
        lock (gate)
        {
            var five = result.FiveHour;
            var weekly = result.Weekly;
            snapshot = snapshot with
            {
                FiveHourRemainingPercent = five is null ? snapshot.FiveHourRemainingPercent : MergeRemaining(snapshot.FiveHourRemainingPercent, snapshot.FiveHourResetsAt, five),
                FiveHourResetsAt = five?.ResetsAt ?? snapshot.FiveHourResetsAt,
                WeeklyRemainingPercent = weekly is null ? snapshot.WeeklyRemainingPercent : MergeRemaining(snapshot.WeeklyRemainingPercent, snapshot.WeeklyResetsAt, weekly),
                WeeklyResetsAt = weekly?.ResetsAt ?? snapshot.WeeklyResetsAt,
                LastUpdatedAt = DateTimeOffset.Now,
                Freshness = Freshness.Fresh
            };
            lastRateLimitUpdate = DateTimeOffset.Now;
        }
        Publish();
    }

    private void OnContext(ContextUsage usage)
    {
        lock (gate)
        {
            snapshot = snapshot with
            {
                ContextRemainingPercent = usage.RemainingPercent,
                ActiveThreadId = usage.ThreadId,
                CodyState = usage.CodyState,
                CodyStateUpdatedAt = usage.CodyStateUpdatedAt,
                AgentPulse = usage.AgentPulse,
                AgentPulseReason = usage.AgentPulseReason,
                LastUpdatedAt = DateTimeOffset.Now,
                Freshness = Freshness.Fresh
            };
            lastContextUpdate = usage.UpdatedAt;
        }
        if (usage.RateLimits is not null) Merge(usage.RateLimits); else Publish();
    }

    private async Task FreshnessLoopAsync(CancellationToken cancellationToken)
    {
        while (!cancellationToken.IsCancellationRequested)
        {
            await Task.Delay(TimeSpan.FromSeconds(1), cancellationToken);
            try { sessionMonitor.RefreshActive(); } catch { }
            lock (gate)
            {
                if (snapshot.CodyStateUpdatedAt is DateTimeOffset stateUpdatedAt)
                {
                    var visibleState = CodyStatePolicy.VisibleState(snapshot.CodyState, stateUpdatedAt, DateTimeOffset.Now);
                    if (visibleState != snapshot.CodyState)
                        snapshot = snapshot with { CodyState = visibleState, CodyStateUpdatedAt = DateTimeOffset.Now };
                }
                var newest = new[] { lastRateLimitUpdate, lastContextUpdate }.Where(x => x is not null).Max();
                if (newest is not null && DateTimeOffset.Now - newest > TimeSpan.FromSeconds(90) && snapshot.Freshness == Freshness.Fresh)
                    snapshot = snapshot with { Freshness = Freshness.Delayed };
            }
            Publish();
        }
    }

    private static int MergeRemaining(int? current, DateTimeOffset? currentReset, RateLimitWindow incoming) =>
        current is not null && currentReset is not null && incoming.ResetsAt is not null && Math.Abs((currentReset.Value - incoming.ResetsAt.Value).TotalSeconds) < 1
            ? Math.Min(current.Value, incoming.RemainingPercent) : incoming.RemainingPercent;
    private static bool HasAnyValue(UsageSnapshot value) => value.ContextRemainingPercent is not null || value.FiveHourRemainingPercent is not null || value.WeeklyRemainingPercent is not null;
    private void Publish() { UsageSnapshot value; lock (gate) value = snapshot; Updated?.Invoke(value); }

    public async ValueTask DisposeAsync()
    {
        lifetime?.Cancel(); lifetime?.Dispose(); lifetime = null;
        sessionMonitor.Dispose();
        await client.DisposeAsync();
    }
}
